File: java/rocksjni/metrics_scanner.cc
//
// Created by jie.huang on 16/11/15.
//

#include <stdio.h>
#include <stdlib.h>
#include <jni.h>
#include <string>

#include "include/org_rocksdb_MetricsScanner.h"
#include "rocksjni/portal.h"
#include "rocksdb/metrics_scanner.h"

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    start
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_enableLog
        (JNIEnv *env, jobject jobj, jlong handle, jboolean enableLog) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->enableLog = enableLog;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    start
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_enableProfiler
        (JNIEnv *env, jobject jobj, jlong handle, jboolean enableProfiler) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->enableProfiler = enableProfiler;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    metricType
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_metricType
        (JNIEnv *env, jobject jobj, jlong handle, jbyte metricType) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->metric_type = metricType;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    groupByLimit
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_groupByLimit
        (JNIEnv *env, jobject jobj, jlong handle, jint groupByLimit) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->groupByLimit = groupByLimit;
}

void Java_org_rocksdb_MetricsScanner_minTagValueLen
        (JNIEnv *env, jobject jobj, jlong handle, jbyte minTagValueLen) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->minTagValueLen = minTagValueLen;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    start
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_metric
        (JNIEnv *env, jobject jobj, jlong handle, jint metric) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->metric = metric;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    startHour
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_start
        (JNIEnv *env, jobject jobj, jlong handle, jint start) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->start = start;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    endHour
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_end
        (JNIEnv *env, jobject jobj, jlong handle, jint end) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->end = end;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    maxPointCount
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_maxPointCount
        (JNIEnv *env, jobject jobj, jlong handle, jint maxPointCount) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->pointCount = maxPointCount;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    doScan
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_next
        (JNIEnv *env, jobject jobj, jlong handle) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->next();
}


/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    doSearch
 * Signature: (JI)V
 */
void Java_org_rocksdb_MetricsScanner_doSearch
        (JNIEnv *env, jobject jobj, jlong handle) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    scanner->doSearch();
}


/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    hasNextBaseTime
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_MetricsScanner_hasNextBaseTime
        (JNIEnv *env, jobject jobj, jlong handle, jbyte baseTime, jlong columnFamilyHandle) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    auto column_family_handle = reinterpret_cast<rocksdb::ColumnFamilyHandle *>(columnFamilyHandle);
    return scanner->hasNextBaseTime(baseTime, column_family_handle);
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    hasNext
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_MetricsScanner_hasNext
        (JNIEnv *env, jobject jobj, jlong handle) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    return scanner->hasNext();
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    getCurrentHour
 * Signature: (J)I
 */
jint Java_org_rocksdb_MetricsScanner_getCurrentBaseTime
        (JNIEnv *env, jobject jobj, jlong handle) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    return scanner->getCurrentBaseTime();
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    getResultSet
 * Signature: (J)[B
 */
jbyteArray Java_org_rocksdb_MetricsScanner_getResultSet
        (JNIEnv *env, jobject jobj, jlong handle) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    rocksdb::Slice value_slice = scanner->getResultSet();
    jsize size = static_cast<jsize>(value_slice.size());
    jbyteArray jkeyValue = env->NewByteArray(size);
    env->SetByteArrayRegion(jkeyValue, 0, size,
                            reinterpret_cast<const jbyte *>(value_slice.data()));
    return jkeyValue;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    getGroupBy
 * Signature: (J)[B
 */
jbyteArray Java_org_rocksdb_MetricsScanner_getGroupBy
        (JNIEnv *env, jobject jobj, jlong handle) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    rocksdb::Slice value_slice = scanner->getGroupBy();
    jsize size = static_cast<jsize>(value_slice.size());
    jbyteArray jkeyValue = env->NewByteArray(size);
    env->SetByteArrayRegion(jkeyValue, 0, size,
                            reinterpret_cast<const jbyte *>(value_slice.data()));
    return jkeyValue;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    getStat
 * Signature: (J)[B
 */
jbyteArray Java_org_rocksdb_MetricsScanner_getStat
        (JNIEnv *env, jobject jobj, jlong handle) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    rocksdb::Slice value_slice = scanner->getStat();
    jsize size = static_cast<jsize>(value_slice.size());
    jbyteArray jkeyValue = env->NewByteArray(size);
    env->SetByteArrayRegion(jkeyValue, 0, size,
                            reinterpret_cast<const jbyte *>(value_slice.data()));
    return jkeyValue;
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    setTagFilters
 * Signature: (JI[B)V
 */
void Java_org_rocksdb_MetricsScanner_setTagFilters
        (JNIEnv *env, jobject jobj, jlong handler, jint jtarget_len, jbyteArray jtarget) {
    auto *scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handler);

    jbyte *target = env->GetByteArrayElements(jtarget, 0);
    rocksdb::Slice target_slice(
            reinterpret_cast<char *>(target), jtarget_len);

    scanner->setTagFilter(target_slice);

    env->ReleaseByteArrayElements(jtarget, target, JNI_ABORT);
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    setGroupBy
 * Signature: (JI[B)V
 */
void Java_org_rocksdb_MetricsScanner_setGroupBy
        (JNIEnv *env, jobject jobj, jlong handler, jint jtarget_len, jbyteArray jtarget) {
    auto *scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handler);

    jbyte *target = env->GetByteArrayElements(jtarget, 0);
    rocksdb::Slice target_slice(
            reinterpret_cast<char *>(target), jtarget_len);

    scanner->setGroupBy(target_slice);

    env->ReleaseByteArrayElements(jtarget, target, JNI_ABORT);
}

/*
 * Class:     org_rocksdb_MetricsScanner
 * Method:    disposeInternal
 * Signature: (J)V
 */
void Java_org_rocksdb_MetricsScanner_disposeInternal
        (JNIEnv *env, jobject jobj, jlong handle) {
    auto scanner = reinterpret_cast<rocksdb::MetricsScanner *>(handle);
    delete scanner;
}

File: java/src/main/java/org/rocksdb/MetricsScanner.java
package org.rocksdb;

/**
 * @author jie.huang
 *         Date: 16/11/15
 *         Time: 下午7:10
 */
public class MetricsScanner extends RocksObject {
    private RocksDB rocksDB;

    public MetricsScanner(RocksDB rocksDB, long nativeHandle) {
        super(nativeHandle);
        this.rocksDB = rocksDB;
    }

    public void setMaxPointCount(int maxPointCount) throws RocksDBException {
        if (maxPointCount <= 0) {
            throw new RocksDBException("Max point count must be > 0.");
        }
        maxPointCount(nativeHandle_, maxPointCount);
    }

    public void setMetricType(byte metricType) throws RocksDBException {
        if (metricType <= 0) {
            throw new RocksDBException("Metric type cannot be empty.");
        }
        metricType(nativeHandle_, metricType);
    }

    public void setMetric(int metric) throws RocksDBException {
        if (metric == -1) {
            throw new RocksDBException("Metric cannot be empty.");
        }
        metric(nativeHandle_, metric);
    }

    public void setEnableLog(boolean enableLog) {
        enableLog(nativeHandle_, enableLog);
    }

    public void setEnableProfiler(boolean enableProfiler) {
        enableProfiler(nativeHandle_, enableProfiler);
    }

    public void setRange(int start, int end) throws RocksDBException {
        if (start == -1) {
            throw new RocksDBException("Start slot cannot be empty.");
        }
        if (end == -1) {
            throw new RocksDBException("End slot cannot be empty.");
        }
        start(nativeHandle_, start);
        end(nativeHandle_, end);
    }

    public void setMinTagValueLen(int minTagValueLen) {
        minTagValueLen(nativeHandle_, (byte) minTagValueLen);
    }

    public void setTagFilters(byte[] tagFilters) {
        if (tagFilters != null && tagFilters.length > 0) {
            setTagFilters(nativeHandle_, tagFilters.length, tagFilters);
        }
    }

    public void setGroupBy(byte[] groupBy) {
        if (groupBy != null && groupBy.length > 0) {
            setGroupBy(nativeHandle_, groupBy.length, groupBy);
        }
    }

    public void next() {
        next(nativeHandle_);
    }

    public boolean hasNextBaseTime(byte baseTime, ColumnFamilyHandle familyHandle) {
        return hasNextBaseTime(nativeHandle_, baseTime, familyHandle.nativeHandle_);
    }

    public boolean hasNext() {
        return hasNext(nativeHandle_);
    }

    public int getCurrentBaseTime() {
        return getCurrentBaseTime(nativeHandle_);
    }

    public byte[] getResultSet() {
        return getResultSet(nativeHandle_);
    }

    public byte[] getGroupBy() {
        return getGroupBy(nativeHandle_);
    }

    public byte[] getStat() {
        return getStat(nativeHandle_);
    }

    public void close() {
        super.close();
    }

    @Override
    protected void disposeInternal() {
        if (rocksDB.isOwningHandle()) {
            disposeInternal(nativeHandle_);
        }
    }

    private native void enableLog(long handler, boolean enable);

    private native void enableProfiler(long handler, boolean enable);

    private native void maxPointCount(long handle, int maxPointCount);

    private native void metric(long handle, int metric);

    private native void metricType(long handle, byte metricType);

    private native void minTagValueLen(long handle, byte minTagValueLen);

    private native void start(long handle, int start);

    private native void end(long handle, int end);

    private native void next(long handle);

    private native boolean hasNextBaseTime(long handle, byte baseTime, long columnFamilyHandle);

    private native boolean hasNext(long handle);

    private native int getCurrentBaseTime(long handle);

    private native byte[] getResultSet(long handle);

    private native byte[] getGroupBy(long handle);

    private native byte[] getStat(long handle);

    private native void setTagFilters(long handle, int len, byte[] tagFilters);

    private native void setGroupBy(long handle, int len, byte[] groupBy);

    @Override
    protected final native void disposeInternal(final long handle);
}
